### PR TITLE
bpo-31373: fix undefined floating-point demotions

### PR DESCRIPTION
**Pull request by benjaminp**
https://bugs.python.org/issue31373

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2017-09-06-15-25-59.bpo-31373.dC4jd4.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Fix several possible instances of undefined behavior due to floating-point
+demotions.
+
+# Write your Misc/NEWS entry below.  It should be a simple ReST paragraph. #
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2017-09-06-15-25-59.bpo-31373.dC4jd4.rst]: You are not supposed to commit this part :-)

File: Python/getargs.c [hunk 2]
```diff
@@ -858,6 +859,10 @@ convertsimple(PyObject *arg, const char **p_format, va_list *p_va, int flags,
         double dval = PyFloat_AsDouble(arg);
         if (PyErr_Occurred())
             RETURN_ERR_OCCURRED;
+        else if (dval > FLT_MAX)
+            *p = (float)INFINITY;
+        else if (dval < -FLT_MAX)
```
Comment (vstinner) [inline Python/getargs.c:864]: I would prefer to use FLT_MIN.

File: Objects/floatobject.c [hunk 3]
```diff
@@ -2233,21 +2233,23 @@ _PyFloat_Pack4(double x, unsigned char *p, int le)
...
     else {
-        float y = (float)x;
-        const unsigned char *s = (unsigned char*)&y;
         int i, incr = 1;
 
-        if (Py_IS_INFINITY(y) && !Py_IS_INFINITY(x))
+        if (fabs(x) > FLT_MAX && !Py_IS_INFINITY(x))
             goto Overflow;
 
+        unsigned char s[sizeof(float)];
```
Comment (vstinner) [inline Objects/floatobject.c:2241]: I would be more confident with an assert(sizeof(float) == 4); just in case.

File: Objects/floatobject.c [hunk 4]
```diff
@@ -2233,21 +2233,23 @@ _PyFloat_Pack4(double x, unsigned char *p, int le)
 
     }
     else {
-        float y = (float)x;
-        const unsigned char *s = (unsigned char*)&y;
         int i, incr = 1;
 
-        if (Py_IS_INFINITY(y) && !Py_IS_INFINITY(x))
```
Comment (vstinner) [inline Objects/floatobject.c:2240]: I don't understand why you remove Py_IS_INFINITY(y).

File: Include/pymath.h [hunk 5]
```diff
@@ -215,4 +215,16 @@ PyAPI_FUNC(void) _Py_set_387controlword(unsigned short);
 					 (X) == -Py_HUGE_VAL))
 #endif
 
+/* Return whether integral type *type* is signed or not. */
+#define _Py_IntegralSigned(type) ((type)(-1) < 0)
```
Comment (vstinner) [inline Include/pymath.h]: I suggest to put "Type" in the macro names. Maybe _Py_IsTypeSigned(type), _Py_TypeMax(), etc.?

File: Include/pymath.h [hunk 6]
```diff
@@ -215,4 +215,16 @@ PyAPI_FUNC(void) _Py_set_387controlword(unsigned short);
...
+/* Return the maximum value of integral type *type*. */
+#define _Py_IntegralMax(type) ((_Py_IntegralSigned(type)) ? (((((type)1 << (sizeof(type)*8 - 2)) - 1) << 1) + 1) : ~(type)0)
+/* Return the minimum value of integral type *type*. */
+#define _Py_IntegralMin(type) ((_Py_IntegralSigned(type)) ? -_Py_IntegralMax(type) - 1 : 0)
+/* Check whether *v* is in the range of integral type *type*. This is most
+ * useful if *v* is floating-point, since demoting a floating-point *v* to an
+ * integral type that cannot represent *v*'s integral part is undefined
+ * behavior. */
+#define _Py_IntegralInRange(type, v) (_Py_IntegralMin(type) <= v && v <= _Py_IntegralMax(type))
```
Comment (vstinner) [inline Include/pymath.h]: _Py_IsInTypeRange(type, v)?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Python/getargs.c (hunk 2 above)
Comment (benjaminp) [reply Python/getargs.c:864]: `FLT_MIN` is the smallest nonnegative float not what we want here.

File: Objects/floatobject.c (hunk 4 above)
Comment (benjaminp) [reply Objects/floatobject.c:2240]: `y` is no longer defined at this point.

File: Objects/floatobject.c (hunk 3 above)
Comment (benjaminp) [reply Objects/floatobject.c:2241]: This code only executes if floats are in IEE754 format and are therefore 4 bytes.

File: Misc/NEWS.d/next/Core and Builtins/2017-09-06-15-25-59.bpo-31373.dC4jd4.rst (hunk 1 above)
Comment (benjaminp) [reply Misc/NEWS.d/next/Core and Builtins/2017-09-06-15-25-59.bpo-31373.dC4jd4.rst]: hmm, blurb hiccup I guess

File: Include/pymath.h (hunk 5 above)
Comment (benjaminp) [reply Include/pymath.h]: sg

File: Python/getargs.c (hunk 2 above)
Comment (benjaminp) [reply Python/getargs.c]: `FLT_MIN` is the smallest nonnegative representable float, so not what we want here.

**Pull request merged**